File: org.kevoree.modeling.cpp.microframework/src/main/cpp/org/kevoree/modeling/api/json/Lexer.cpp
#include "Lexer.h"

#include <algorithm>
#include <string> 
#include <sstream>

Lexer::Lexer(istream &inputstream)
{
	if(inputstream)
	{		
		inputstream.seekg (0, inputstream.end);
		length = inputstream.tellg();
		inputstream.seekg (0, inputstream.beg);
		bytes = new char [length];
		//std::cout << "Reading " << length << " characters... "  << endl;
		inputstream.read (bytes,length);	
	}
	
	finish_token = new Token(LexerType(END_OF_FILE),"");
	index =0;
}


bool Lexer::isSpace(char c){
 return c == ' ' || c == '\r' || c == '\n' || c == '\t';		
}



char Lexer::nextChar()
 {
	return bytes[index++];   
}

char Lexer::peekChar()
 {
	return bytes[index];  
}

bool Lexer::isDone()
 {
	return  index >= length;
}


bool Lexer::isBooleanLetter(char c){
	if(c  == 'f' || c == 'a' || c == 'l' || c == 's' || c == 'e' || c == 't' || c== 'r' || c== 'u'){
		return true;
	}else {
		return false;
	}
}

bool Lexer::isDigit(char c)
{
	if(c  == '0' || c == '1' || c == '2' || c == '3' || c == '4' || c== '5' || c== '6' || c== '7' || c== '8' || c=='9'){
		return true;
	}else 
	{
		return false;
	}
}


 bool Lexer::isValueLetter(char c)  
 {
  return c == '-' || c == '+' || c == '.' || isDigit(c) || isBooleanLetter(c);
 }


Token Lexer::nextToken()  {
	
	    if (isDone()) {
            return *finish_token;
        }
        int  tokenType = LexerType(END_OF_FILE);
        
        char c = nextChar();

        std::ostringstream currentValue;
        string jsonValue ="";
        while (!isDone() && isSpace(c)) {
            c = nextChar();   
        }
        
        if ('"' == c) {
            tokenType = LexerType(VALUE);
            if (! isDone()) {
                c = nextChar();
                while (index < length && c != '"') {
                  currentValue << c;
                    if (c == '\\' && index < length) {
                        c = nextChar();
                        currentValue << c;
                    }
                    c = nextChar();
                }
                jsonValue = currentValue.str();
            } else {
                   throw  "Unterminated string";
            }
        }else if ('{' == c) {
            tokenType = LexerType(LEFT_BRACE);
        } else if ('}' == c) {
            tokenType = LexerType(RIGHT_BRACE);
        } else if ('[' == c) {
            tokenType = LexerType(LEFT_BRACKET);
        } else if (']' == c) {
            tokenType = LexerType(RIGHT_BRACKET);
        } else if (':' == c) {
            tokenType = LexerType(COLON);
        } else if (',' == c) {
            tokenType = LexerType(COMMA);
        } else if (! isDone()) {
            while (isValueLetter(c)) {
            currentValue << c;
                if (! isValueLetter(peekChar())) {
                    break;
                } else {
                    c = nextChar();
                }
            }
            string v = currentValue.str();
            std::transform(v.begin(), v.end(), v.begin(), ::tolower);
            if (v.compare("true")) {
                jsonValue = "true";
            } else if (v.compare("false")) {
                jsonValue = "false";
            } else {
                jsonValue = v;
            }
            tokenType = LexerType(VALUE);
        } else {
            tokenType = LexerType(END_OF_FILE);
        }
     
              return Token(tokenType, jsonValue);
}

File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/json/Lexer.ts
///<reference path="../util/Converters.ts"/>
///<reference path="../../../../../java/util/JUHashSet.ts"/>

class Lexer {

  private bytes: number[] = null;
  private EOF: JsonToken = null;
  private BOOLEAN_LETTERS: JUHashSet<string> = null;
  private DIGIT: JUHashSet<string> = null;
  private index: number = 0;
  private static DEFAULT_BUFFER_SIZE: number = 1024 * 4;

  constructor(payload: string) {
    this.bytes = payload.getBytes();
    this.EOF = new JsonToken(Type.EOF, null);
  }

  public isSpace(c: string): boolean {
    return c == ' ' || c == '\r' || c == '\n' || c == '\t';
  }

  private nextChar(): string {
    return new Converters().toChar(this.bytes[this.index++]);
  }

  private peekChar(): string {
    return new Converters().toChar(this.bytes[this.index]);
  }

  private isDone(): boolean {
    return this.index >= this.bytes.length;
  }

  private isBooleanLetter(c: string): boolean {
    if (this.BOOLEAN_LETTERS == null) {
      this.BOOLEAN_LETTERS = new JUHashSet<string>();
      this.BOOLEAN_LETTERS.add('f');
      this.BOOLEAN_LETTERS.add('a');
      this.BOOLEAN_LETTERS.add('l');
      this.BOOLEAN_LETTERS.add('s');
      this.BOOLEAN_LETTERS.add('e');
      this.BOOLEAN_LETTERS.add('t');
      this.BOOLEAN_LETTERS.add('r');
      this.BOOLEAN_LETTERS.add('u');
    }
    return this.BOOLEAN_LETTERS.contains(c);
  }

  private isDigit(c: string): boolean {
    if (this.DIGIT == null) {
      this.DIGIT = new JUHashSet<string>();
      this.DIGIT.add('0');
      this.DIGIT.add('1');
      this.DIGIT.add('2');
      this.DIGIT.add('3');
      this.DIGIT.add('4');
      this.DIGIT.add('5');
      this.DIGIT.add('6');
      this.DIGIT.add('7');
      this.DIGIT.add('8');
      this.DIGIT.add('9');
    }
    return this.DIGIT.contains(c);
  }

  private isValueLetter(c: string): boolean {
    return c == '-' || c == '+' || c == '.' || this.isDigit(c) || this.isBooleanLetter(c);
  }

  public nextToken(): JsonToken {
    if (this.isDone()) {
      return this.EOF;
    }
    var tokenType: Type = Type.EOF;
    var c: string = this.nextChar();
    var currentValue: StringBuilder = new StringBuilder();
    var jsonValue: any = null;
    while (!this.isDone() && this.isSpace(c)){
      c = this.nextChar();
    }
    if ('"' == c) {
      tokenType = Type.VALUE;
      if (!this.isDone()) {
        c = this.nextChar();
        while (this.index < this.bytes.length && c != '"'){
          currentValue.append(c);
          if (c == '\\' && this.index < this.bytes.length) {
            c = this.nextChar();
            currentValue.append(c);
          }
          c = this.nextChar();
        }
        jsonValue = currentValue.toString();
      }
    } else {
      if ('{' == c) {
        tokenType = Type.LEFT_BRACE;
      } else {
        if ('}' == c) {
          tokenType = Type.RIGHT_BRACE;
        } else {
          if ('[' == c) {
            tokenType = Type.LEFT_BRACKET;
          } else {
            if (']' == c) {
              tokenType = Type.RIGHT_BRACKET;
            } else {
              if (':' == c) {
                tokenType = Type.COLON;
              } else {
                if (',' == c) {
                  tokenType = Type.COMMA;
                } else {
                  if (!this.isDone()) {
                    while (this.isValueLetter(c)){
                      currentValue.append(c);
                      if (!this.isValueLetter(this.peekChar())) {
                        break;
                      } else {
                        c = this.nextChar();
                      }
                    }
                    var v: string = currentValue.toString();
                    if ("true".equals(v.toLowerCase())) {
                      jsonValue = true;
                    } else {
                      if ("false".equals(v.toLowerCase())) {
                        jsonValue = false;
                      } else {
                        jsonValue = v.toLowerCase();
                      }
                    }
                    tokenType = Type.VALUE;
                  } else {
                    tokenType = Type.EOF;
                  }
                }
              }
            }
          }
        }
      }
    }
    return new JsonToken(tokenType, jsonValue);
  }

}

